File: src/kernel/pes/PEMux.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Kernel.h>

#include "pes/PEMux.h"
#include "pes/VPEManager.h"
#include "TCU.h"
#include "Platform.h"
#include "SyscallHandler.h"

namespace kernel {

PEMux::PEMux(peid_t pe)
    : _pe(new PEObject(pe, EP_COUNT - m3::TCU::FIRST_USER_EP)),
      _caps(VPE::INVALID_ID),
      _vpes(),
      _rbufs_size(),
      _mem_base(),
      _eps(),
      _upcqueue(desc()) {
    for(epid_t ep = 0; ep < m3::TCU::FIRST_USER_EP; ++ep)
        _eps.set(ep);

#if defined(__gem5__)
    if(Platform::pe(pe).supports_pemux()) {
        // configure send EP
        TCU::config_remote_ep(0, pe, m3::TCU::KPEX_SEP, [this](m3::TCU::reg_t *ep_regs) {
            TCU::config_send(ep_regs, m3::KIF::PEMUX_VPE_ID, m3::ptr_to_label(this),
                             Platform::kernel_pe(), TCU::PEX_REP, KPEX_RBUF_ORDER, 1);
        });

        // configure receive EP
        uintptr_t rbuf = PEMUX_RBUF_SPACE;
        TCU::config_remote_ep(0, pe, m3::TCU::KPEX_REP, [rbuf](m3::TCU::reg_t *ep_regs) {
            TCU::config_recv(ep_regs, m3::KIF::PEMUX_VPE_ID, rbuf,
                             KPEX_RBUF_ORDER, KPEX_RBUF_ORDER, m3::TCU::NO_REPLIES);
        });
        rbuf += KPEX_RBUF_SIZE;

        // configure upcall receive EP
        TCU::config_remote_ep(0, pe, m3::TCU::PEXUP_REP, [rbuf](m3::TCU::reg_t *ep_regs) {
            TCU::config_recv(ep_regs, m3::KIF::PEMUX_VPE_ID, rbuf,
                             PEXUP_RBUF_ORDER, PEXUP_RBUF_ORDER, m3::TCU::PEXUP_RPLEP);
        });
    }
    #endif
}

void PEMux::handle_call(const m3::TCU::Message *msg) {
    auto req = reinterpret_cast<const m3::KIF::PEXCalls::Exit*>(msg->data);
    capsel_t vpe = req->vpe_sel;
    int exitcode = req->code;

    KLOG(PEXC, "PEMux[" << peid() << "] got exit(vpe=" << vpe << ", code=" << exitcode << ")");

    auto vpecap = static_cast<VPECapability*>(_caps.get(vpe, Capability::VIRTPE));
    if(vpecap != nullptr) {
        vpecap->obj->_flags |= VPE::F_STOPPED;
        vpecap->obj->stop_app(exitcode, true);
    }

    // give credits back
    TCU::reply(TCU::PEX_REP, nullptr, 0, msg);
}

void PEMux::add_vpe(VPECapability *vpe) {
    assert(_vpes == 0);
    _caps.obtain(vpe->obj->id(), vpe);
    _vpes++;
}

void PEMux::remove_vpe(UNUSED VPE *vpe) {
    // has already been revoked
    assert(_caps.get(vpe->id(), Capability::VIRTPE) == nullptr);
    _vpes--;
    _rbufs_size = 0;
    _mem_base = 0;
}

epid_t PEMux::find_eps(uint count) const {
    uint bit, start = _eps.first_clear();
    for(bit = start; bit < start + count && bit < EP_COUNT; ++bit) {
        if(_eps.is_set(bit))
            start = bit + 1;
    }
    if(bit != start + count)
        return EP_COUNT;
    return start;
}

bool PEMux::eps_free(epid_t start, uint count) const {
    for(epid_t ep = start; ep < start + count; ++ep) {
        if(_eps.is_set(ep))
            return false;
    }
    return true;
}

void PEMux::alloc_eps(epid_t first, uint count) {
    KLOG(EPS, "PEMux[" << peid() << "] allocating EPs " << first << " .. " << (first + count - 1));
    for(uint bit = first; bit < first + count; ++bit)
        _eps.set(bit, true);
}

void PEMux::free_eps(epid_t first, uint count) {
    KLOG(EPS, "PEMux[" << peid() << "] freeing EPs " << first << ".." << (first + count - 1));

    for(epid_t ep = first; ep < first + count; ++ep) {
        assert(_eps.is_set(ep));
        _eps.clear(ep);
    }
}

m3::Errors::Code PEMux::map(vpeid_t vpe, goff_t virt, gaddr_t phys, uint pages, uint perm) {
    m3::KIF::PEXUpcalls::Map req;
    req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::MAP);
    req.vpe_sel = vpe;
    req.virt = virt;
    req.phys = phys;
    req.pages = pages;
    req.perm = static_cast<xfer_t>(perm);

    KLOG(PEXC, "PEMux[" << peid() << "] sending map(vpe=" << req.vpe_sel
        << ", virt=" << m3::fmt((void*)req.virt, "p") << ", phys=" << m3::fmt((void*)req.phys, "p")
        << ", pages=" << req.pages << ", perm=" << req.perm << ")");

    return upcall(&req, sizeof(req));
}

m3::Errors::Code PEMux::vpe_ctrl(VPE *vpe, m3::KIF::PEXUpcalls::VPEOp ctrl) {
    static const char *ctrls[] = {
        "INIT", "START", "STOP"
    };

    m3::KIF::PEXUpcalls::VPECtrl req;
    req.opcode = static_cast<xfer_t>(m3::KIF::PEXUpcalls::VPE_CTRL);
    req.vpe_sel = vpe->id();
    req.vpe_op = ctrl;
    req.eps_start = vpe->eps_start();

    KLOG(PEXC, "PEMux[" << peid() << "] sending VPECtrl(vpe="
        << req.vpe_sel << ", ctrl=" << ctrls[req.vpe_op] << ")");

    return upcall(&req, sizeof(req));
}

m3::Errors::Code PEMux::upcall(void *req, size_t size) {
    // send upcall
    event_t event = _upcqueue.send(m3::TCU::PEXUP_REP, 0, req, size, false);
    m3::ThreadManager::get().wait_for(event);

    // wait for reply
    auto reply_msg = reinterpret_cast<const m3::TCU::Message*>(m3::ThreadManager::get().get_current_msg());
    auto reply = reinterpret_cast<const m3::KIF::DefaultReply*>(reply_msg->data);
    return static_cast<m3::Errors::Code>(reply->error);
}

m3::Errors::Code PEMux::invalidate_ep(vpeid_t vpe, epid_t ep, bool force) {
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = invalid");

    uint32_t unread_mask;
    m3::Errors::Code res = TCU::inval_ep_remote(vpe, peid(), ep, force, &unread_mask);
    if(res != m3::Errors::NONE)
        return res;

    if(unread_mask != 0) {
        m3::KIF::PEXUpcalls::RemMsgs req;
        req.vpe_sel = vpe;
        req.unread_mask = unread_mask;
        return upcall(&req, sizeof(req));
    }
    else
        return m3::Errors::NONE;
}

m3::Errors::Code PEMux::config_rcv_ep(epid_t ep, vpeid_t vpe, epid_t rpleps, RGateObject &obj) {
    assert(obj.activated());
    // it needs to be in the receive buffer space
    const goff_t addr = Platform::def_recvbuf(peid());
    const size_t size = Platform::pe(peid()).has_virtmem() ? RECVBUF_SIZE : RECVBUF_SIZE_SPM;
    // def_recvbuf() == 0 means that we do not validate it
    if(addr && (obj.addr < addr || obj.addr > addr + size || obj.addr + obj.size() > addr + size))
        return m3::Errors::INV_ARGS;
    if(obj.addr < addr + _rbufs_size)
        return m3::Errors::INV_ARGS;

    vpeid_t ep_vpe = Platform::is_shared(peid()) ? vpe : VPE::INVALID_ID;
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = "
        "RGate[vpe=" << ep_vpe << ", addr=#" << m3::fmt(obj.addr, "x")
        << ", order=" << obj.order
        << ", msgorder=" << obj.msgorder
        << ", replyeps=" << rpleps
        << "]");

    TCU::config_remote_ep(vpe, peid(), ep, [this, &obj, rpleps, ep_vpe](m3::TCU::reg_t *ep_regs) {
        TCU::config_recv(ep_regs, ep_vpe, rbuf_base() + obj.addr,
                         obj.order, obj.msgorder, rpleps);
    });

    m3::ThreadManager::get().notify(reinterpret_cast<event_t>(&obj));
    return m3::Errors::NONE;
}

m3::Errors::Code PEMux::config_snd_ep(epid_t ep, vpeid_t vpe, SGateObject &obj) {
    assert(obj.rgate->addr != 0);
    if(obj.activated)
        return m3::Errors::EXISTS;

    vpeid_t ep_vpe = Platform::is_shared(peid()) ? vpe : VPE::INVALID_ID;
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = "
        "Send[vpe=" << ep_vpe << ", pe=" << obj.rgate->pe
        << ", ep=" << obj.rgate->ep
        << ", label=#" << m3::fmt(obj.label, "x")
        << ", msgsize=" << obj.rgate->msgorder
        << ", crd=#" << m3::fmt(obj.credits, "x")
        << "]");

    obj.activated = true;

    TCU::config_remote_ep(vpe, peid(), ep, [&obj, ep_vpe](m3::TCU::reg_t *ep_regs) {
        TCU::config_send(ep_regs, ep_vpe, obj.label, obj.rgate->pe, obj.rgate->ep,
                         obj.rgate->msgorder, obj.credits);
    });
    return m3::Errors::NONE;
}

m3::Errors::Code PEMux::config_mem_ep(epid_t ep, vpeid_t vpe, const MGateObject &obj, goff_t off) {
    if(off >= obj.size || obj.addr + off < off)
        return m3::Errors::INV_ARGS;

    vpeid_t ep_vpe = Platform::is_shared(peid()) ? vpe : VPE::INVALID_ID;
    KLOG(EPS, "PE" << peid() << ":EP" << ep << " = "
        "Mem [vpe=" << ep_vpe << ", pe=" << obj.pe
        << ", addr=#" << m3::fmt(obj.addr + off, "x")
        << ", size=#" << m3::fmt(obj.size - off, "x")
        << ", perms=#" << m3::fmt(obj.perms, "x")
        << "]");

    TCU::config_remote_ep(vpe, peid(), ep, [&obj, ep_vpe, off](m3::TCU::reg_t *ep_regs) {
        TCU::config_mem(ep_regs, ep_vpe, obj.pe, obj.vpe,
                        obj.addr + off, obj.size - off, obj.perms);
    });
    return m3::Errors::NONE;
}

}

File: src/kernel/src/pes/pemux.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base::col::{BitVec, Vec};
use base::errors::{Code, Error};
use base::goff;
use base::kif;
use base::mem::GlobAddr;
use base::rc::{Rc, SRc, Weak};
use base::mem::{MsgBuf, MsgBufRef};
use base::tcu::{self, EpId, PEId, VPEId};
use core::cmp;

use crate::cap::{EPObject, MGateObject, PEObject, RGateObject, SGateObject};
use crate::ktcu;
use crate::pes::INVAL_ID;
use crate::platform;

pub struct PEMux {
    pe: SRc<PEObject>,
    vpes: Vec<VPEId>,
    #[cfg(target_os = "none")]
    queue: crate::com::SendQueue,
    pmp: Vec<Rc<EPObject>>,
    eps: BitVec,
    #[allow(dead_code)]
    mem_base: goff,
}

impl PEMux {
    pub fn new(pe: PEId) -> Self {
        let pe_obj = PEObject::new(pe, (tcu::AVAIL_EPS - tcu::FIRST_USER_EP) as u32, false);

        // create PMP EPObjects for this PE
        let mut pmp = Vec::new();
        for ep in 0..tcu::PMEM_PROT_EPS as EpId {
            pmp.push(EPObject::new(false, Weak::new(), ep, 0, &pe_obj));
        }

        let mut pemux = PEMux {
            pe: pe_obj,
            vpes: Vec::new(),
            #[cfg(target_os = "none")]
            queue: crate::com::SendQueue::new(pe as u64, pe),
            pmp,
            eps: BitVec::new(tcu::AVAIL_EPS as usize),
            mem_base: 0,
        };

        #[cfg(target_os = "none")]
        pemux.eps.set(0); // first EP is reserved for PEMux's memory region

        for ep in tcu::PMEM_PROT_EPS as EpId..tcu::FIRST_USER_EP {
            pemux.eps.set(ep as usize);
        }

        #[cfg(target_os = "none")]
        if platform::pe_desc(pe).supports_pemux() {
            pemux.init();
        }

        pemux
    }

    pub fn has_vpes(&self) -> bool {
        !self.vpes.is_empty()
    }

    pub fn add_vpe(&mut self, vpe: VPEId) {
        self.vpes.push(vpe);
    }

    pub fn rem_vpe(&mut self, vpe: VPEId) {
        assert!(!self.vpes.is_empty());
        self.vpes.retain(|id| *id != vpe);
    }

    #[cfg(target_os = "none")]
    fn init(&mut self) {
        use base::cfg;

        // configure send EP
        ktcu::config_remote_ep(self.pe_id(), tcu::KPEX_SEP, |regs| {
            ktcu::config_send(
                regs,
                kif::pemux::VPE_ID as VPEId,
                self.pe_id() as tcu::Label,
                platform::kernel_pe(),
                ktcu::KPEX_EP,
                cfg::KPEX_RBUF_ORD,
                1,
            );
        })
        .unwrap();

        // configure receive EP
        let mut rbuf = platform::rbuf_pemux(self.pe_id());
        ktcu::config_remote_ep(self.pe_id(), tcu::KPEX_REP, |regs| {
            ktcu::config_recv(
                regs,
                kif::pemux::VPE_ID as VPEId,
                rbuf,
                cfg::KPEX_RBUF_ORD,
                cfg::KPEX_RBUF_ORD,
                None,
            );
        })
        .unwrap();
        rbuf += 1 << cfg::KPEX_RBUF_ORD;

        // configure upcall EP
        ktcu::config_remote_ep(self.pe_id(), tcu::PEXSIDE_REP, |regs| {
            ktcu::config_recv(
                regs,
                kif::pemux::VPE_ID as VPEId,
                rbuf,
                cfg::PEXUP_RBUF_ORD,
                cfg::PEXUP_RBUF_ORD,
                Some(tcu::PEXSIDE_RPLEP),
            );
        })
        .unwrap();
    }

    pub fn pe(&self) -> &SRc<PEObject> {
        &self.pe
    }

    pub fn pe_id(&self) -> PEId {
        self.pe.pe()
    }

    #[cfg(target_os = "linux")]
    pub fn eps_base(&mut self) -> goff {
        self.mem_base
    }

    pub fn set_mem_base(&mut self, addr: goff) {
        self.mem_base = addr;
    }

    pub fn pmp_ep(&self, ep: EpId) -> &Rc<EPObject> {
        &self.pmp[ep as usize]
    }

    pub fn find_eps(&self, count: u32) -> Result<EpId, Error> {
        // the PMP EPs cannot be allocated
        let mut start = cmp::max(tcu::FIRST_USER_EP as usize, self.eps.first_clear());
        let mut bit = start;
        while bit < start + count as usize && bit < tcu::AVAIL_EPS as usize {
            if self.eps.is_set(bit) {
                start = bit + 1;
            }
            bit += 1;
        }

        if bit != start + count as usize {
            Err(Error::new(Code::NoSpace))
        }
        else {
            Ok(start as EpId)
        }
    }

    pub fn eps_free(&self, start: EpId, count: u32) -> bool {
        for ep in start..start + count as EpId {
            if self.eps.is_set(ep as usize) {
                return false;
            }
        }
        true
    }

    pub fn alloc_eps(&mut self, start: EpId, count: u32) {
        klog!(
            EPS,
            "PEMux[{}] allocating EPS {}..{}",
            self.pe_id(),
            start,
            start as u32 + count - 1
        );
        for bit in start..start + count as EpId {
            assert!(!self.eps.is_set(bit as usize));
            self.eps.set(bit as usize);
        }
    }

    pub fn free_eps(&mut self, start: EpId, count: u32) {
        klog!(
            EPS,
            "PEMux[{}] freeing EPS {}..{}",
            self.pe_id(),
            start,
            start as u32 + count - 1
        );
        for bit in start..start + count as EpId {
            assert!(self.eps.is_set(bit as usize));
            self.eps.clear(bit as usize);
        }
    }

    fn ep_vpe_id(&self, vpe: VPEId) -> VPEId {
        match platform::is_shared(self.pe_id()) {
            true => vpe,
            false => INVAL_ID,
        }
    }

    pub fn config_snd_ep(
        &mut self,
        ep: EpId,
        vpe: VPEId,
        obj: &SRc<SGateObject>,
    ) -> Result<(), Error> {
        let rgate = obj.rgate();
        assert!(rgate.activated());

        klog!(EPS, "PE{}:EP{} = {:?}", self.pe_id(), ep, obj);

        ktcu::config_remote_ep(self.pe_id(), ep, |regs| {
            let vpe = self.ep_vpe_id(vpe);
            let (rpe, rep) = rgate.location().unwrap();
            ktcu::config_send(
                regs,
                vpe,
                obj.label(),
                rpe,
                rep,
                rgate.msg_order(),
                obj.credits(),
            );
        })
    }

    pub fn config_rcv_ep(
        &mut self,
        ep: EpId,
        vpe: VPEId,
        reply_eps: Option<EpId>,
        obj: &SRc<RGateObject>,
    ) -> Result<(), Error> {
        klog!(EPS, "PE{}:EP{} = {:?}", self.pe_id(), ep, obj);

        ktcu::config_remote_ep(self.pe_id(), ep, |regs| {
            let vpe = self.ep_vpe_id(vpe);
            ktcu::config_recv(
                regs,
                vpe,
                obj.addr(),
                obj.order(),
                obj.msg_order(),
                reply_eps,
            );
        })?;

        thread::ThreadManager::get().notify(obj.get_event(), None);
        Ok(())
    }

    pub fn config_mem_ep(
        &mut self,
        ep: EpId,
        vpe: VPEId,
        obj: &SRc<MGateObject>,
        pe_id: PEId,
    ) -> Result<(), Error> {
        if ep < tcu::PMEM_PROT_EPS as EpId {
            klog!(EPS, "PE{}:PMPEP{} = {:?}", self.pe_id(), ep, obj);
        }
        else {
            klog!(EPS, "PE{}:EP{} = {:?}", self.pe_id(), ep, obj);
        }

        ktcu::config_remote_ep(self.pe_id(), ep, |regs| {
            let vpe = self.ep_vpe_id(vpe);
            ktcu::config_mem(
                regs,
                vpe,
                pe_id,
                obj.offset(),
                obj.size() as usize,
                obj.perms(),
            );
        })
    }

    pub fn invalidate_ep(
        &mut self,
        vpe: VPEId,
        ep: EpId,
        force: bool,
        notify: bool,
    ) -> Result<(), Error> {
        klog!(EPS, "PE{}:EP{} = invalid", self.pe_id(), ep);

        let unread = ktcu::invalidate_ep_remote(self.pe_id(), ep, force)?;
        if unread != 0 && notify {
            let mut msg = MsgBuf::borrow_def();
            msg.set(kif::pemux::RemMsgs {
                op: kif::pemux::Sidecalls::REM_MSGS.val as u64,
                vpe_sel: vpe as u64,
                unread_mask: unread as u64,
            });

            self.send_sidecall::<kif::pemux::RemMsgs>(Some(vpe), &msg)
                .map(|_| ())
        }
        else {
            Ok(())
        }
    }

    pub fn invalidate_reply_eps(
        &self,
        recv_pe: PEId,
        recv_ep: EpId,
        send_ep: EpId,
    ) -> Result<(), Error> {
        klog!(
            EPS,
            "PE{}:EP{} = invalid reply EPs at PE{}:EP{}",
            self.pe_id(),
            send_ep,
            recv_pe,
            recv_ep
        );

        ktcu::inv_reply_remote(recv_pe, recv_ep, self.pe_id(), send_ep)
    }
}

#[cfg(target_os = "none")]
impl PEMux {
    pub fn handle_call_async(&mut self, msg: &tcu::Message) {
        use crate::pes::VPEMng;

        let req = msg.get_data::<kif::pemux::Exit>();
        let vpe_id = req.vpe_sel as VPEId;
        let exitcode = req.code as i32;

        klog!(PEXC, "PEMux[{}] received {:?}", self.pe_id(), req);

        if self.vpes.contains(&vpe_id) {
            let vpe = VPEMng::get().vpe(vpe_id).unwrap();
            vpe.stop_app_async(exitcode, true);
        }

        let mut reply = MsgBuf::borrow_def();
        reply.set(kif::DefaultReply { error: 0 });
        ktcu::reply(ktcu::KPEX_EP, &reply, msg).unwrap();
    }

    pub fn vpe_ctrl_async(
        &mut self,
        vpe: VPEId,
        eps_start: EpId,
        ctrl: base::kif::pemux::VPEOp,
    ) -> Result<(), Error> {
        let mut msg = MsgBuf::borrow_def();
        msg.set(kif::pemux::VPECtrl {
            op: kif::pemux::Sidecalls::VPE_CTRL.val as u64,
            vpe_sel: vpe as u64,
            vpe_op: ctrl.val as u64,
            eps_start: eps_start as u64,
        });

        self.send_receive_sidecall_async::<kif::pemux::VPECtrl>(None, msg)
            .map(|_| ())
    }

    pub fn map_async(
        &mut self,
        vpe: VPEId,
        virt: goff,
        glob: GlobAddr,
        pages: usize,
        perm: kif::PageFlags,
    ) -> Result<(), Error> {
        let mut msg = MsgBuf::borrow_def();
        msg.set(kif::pemux::Map {
            op: kif::pemux::Sidecalls::MAP.val as u64,
            vpe_sel: vpe as u64,
            virt: virt as u64,
            global: glob.raw(),
            pages: pages as u64,
            perm: perm.bits() as u64,
        });

        self.send_receive_sidecall_async::<kif::pemux::Map>(Some(vpe), msg)
            .map(|_| ())
    }

    pub fn unmap_async(&mut self, vpe: VPEId, virt: goff, pages: usize) -> Result<(), Error> {
        self.map_async(vpe, virt, GlobAddr::new(0), pages, kif::PageFlags::empty())
    }

    pub fn translate_async(
        &mut self,
        vpe: VPEId,
        virt: goff,
        perm: kif::Perm,
    ) -> Result<GlobAddr, Error> {
        use base::cfg::PAGE_MASK;

        let mut msg = MsgBuf::borrow_def();
        msg.set(kif::pemux::Translate {
            op: kif::pemux::Sidecalls::TRANSLATE.val as u64,
            vpe_sel: vpe as u64,
            virt: virt as u64,
            perm: perm.bits() as u64,
        });

        self.send_receive_sidecall_async::<kif::pemux::Translate>(Some(vpe), msg)
            .map(|reply| GlobAddr::new(reply.val & !(PAGE_MASK as goff)))
    }

    pub fn notify_invalidate(&mut self, vpe: VPEId, ep: EpId) -> Result<(), Error> {
        let mut msg = MsgBuf::borrow_def();
        msg.set(kif::pemux::EpInval {
            op: kif::pemux::Sidecalls::EP_INVAL.val as u64,
            vpe_sel: vpe as u64,
            ep: ep as u64,
        });

        self.send_sidecall::<kif::pemux::EpInval>(Some(vpe), &msg)
            .map(|_| ())
    }

    fn send_sidecall<R: core::fmt::Debug>(
        &mut self,
        vpe: Option<VPEId>,
        req: &MsgBuf,
    ) -> Result<thread::Event, Error> {
        use crate::pes::{State, VPEMng};

        // if the VPE has no app anymore, don't send the notify
        if let Some(id) = vpe {
            if !VPEMng::get()
                .vpe(id)
                .map(|v| v.state() != State::DEAD)
                .unwrap_or(false)
            {
                return Err(Error::new(Code::VPEGone));
            }
        }

        klog!(PEXC, "PEMux[{}] sending {:?}", self.pe_id(), req.get::<R>());

        self.queue.send(tcu::PEXSIDE_REP, 0, req)
    }

    fn send_receive_sidecall_async<R: core::fmt::Debug>(
        &mut self,
        vpe: Option<VPEId>,
        req: MsgBufRef<'_>,
    ) -> Result<&'static kif::pemux::Response, Error> {
        use crate::com::SendQueue;

        let event = self.send_sidecall::<R>(vpe, &req)?;
        drop(req);

        let reply = SendQueue::receive_async(event)?;

        let reply = reply.get_data::<kif::pemux::Response>();
        if reply.error == 0 {
            Ok(reply)
        }
        else {
            Err(Error::new(Code::from(reply.error as u32)))
        }
    }
}

#[cfg(target_os = "linux")]
impl PEMux {
    pub fn update_eps(&mut self) -> Result<(), Error> {
        ktcu::update_eps(self.pe_id(), self.mem_base)
    }

    pub fn vpe_ctrl_async(
        &mut self,
        _vpe: VPEId,
        _eps_start: EpId,
        _ctrl: base::kif::pemux::VPEOp,
    ) -> Result<(), Error> {
        Ok(())
    }

    pub fn map_async(
        &mut self,
        _vpe: VPEId,
        _virt: goff,
        _glob: GlobAddr,
        _pages: usize,
        _perm: kif::PageFlags,
    ) -> Result<(), Error> {
        Ok(())
    }

    pub fn unmap_async(&mut self, _vpe: VPEId, _virt: goff, _pages: usize) -> Result<(), Error> {
        Ok(())
    }

    pub fn notify_invalidate(&mut self, _vpe: VPEId, _ep: EpId) -> Result<(), Error> {
        Ok(())
    }

    fn send_sidecall<R: core::fmt::Debug>(
        &mut self,
        _vpe: Option<VPEId>,
        _req: &MsgBuf,
    ) -> Result<thread::Event, Error> {
        Err(Error::new(Code::NotSup))
    }
}
